SOL-236 Manual Enrollments

muhhshoaib opened the pull request:
@chrisndodge here are the changes that we did in the Manual Enrollments make auditable story.
We are waiting for detailed enrollment to merge into master and then we can update that report and add
the manual enrollment source in the CSV.

- Created new ManualEnrollmentAudit table
- added the reason field in the instructor dashboard before creating manual enrollments.


For reviewers:

SANDBOX: manual-enrollment.m.sandbox.edx.org
JIRA: https://openedx.atlassian.net/browse/SOL-236
- login as 'staff' and see the 'edX Demo Course' which has the Enrollment Report turned on

chrisndodge commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -37,6 +37,12 @@
+      ${_("Enter the reason why the students are Enroll or unEnroll manually.")}
+      ${_("The reason field cannot be left blank.")} </label>
+    <textarea rows="2" id="reason-field-id" name="reason-field" placeholder="${_('Reason')}" spellcheck="false"></textarea>
+  </p>
```
Sorry if this was not clear in the story requirement. I think we just need to make the reason be mandatory if the course is a WhiteLabel course (e.g. course_mode price != 0.0 and mode='honor'). I'm not sure we want to make this change to be global on every course.

chrisndodge commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -43,8 +43,9 @@
-)
+    CourseEnrollment, CourseEnrollmentAllowed, NonExistentCourseError,
+    ManualEnrollmentAudit, UN_ENROLL_TO_ENROLL, ENROLL_TO_UN_ENROLL, ALLOWED_TO_ENROLL_TO_UN_ENROLL, ENROLL_TO_ENROLL,
+    UN_ENROLLED_TO_ALLOWED_TO_ENROLL, UN_ENROLL_TO_UN_ENROLL, ALLOWED_TO_ENROLL_TO_ENROLL)
```
while this doesn't violate pep8/pylint for line length, it seems a bit excessively long. Can you break it up a bit?

chrisndodge commented:
Hi, one general comment. In lms/djangoapps/instructor/views/api.py, I believe we added a capability (feature flagged) to upload a CSV file to create a new account and enroll user. Can you tie that workflow into this audit table as well?

chrisndodge commented on lms/static/coffee/src/instructor_dashboard/membership.coffee:
```diff
@@ -374,12 +375,17 @@ class BatchEnrollment
+      if not @$reason_field.val()
+        @fail_with_error gettext "Reason field should not be left blank."
+        return false
+
```
As mentioned below, I think only paywalled courses should have a mandatory reason filled in

chrisndodge replied:
Also, are there any existing Jasmine tests here that might need to be updated?

chrisndodge commented:
@shnayder @explorerleslie an early heads up on this one. This is to help support auditing reports for our partners to better track when manual enrollments happen (and who did it and why) in paywalled courses. Can you recommend a reviewer. This is still WIP for now.

explorerleslie commented:
@chrisndodge you can work with @cahrens directly on any code reviews that you need. Who's the product owner for this?

chrisndodge commented:
@explorerleslie , @pbaruah is the primary product owner, but this work is in support to a specification that @shnayder originally wrote up

muhhshoaib commented:
@chrisndodge i have updated the detailed enrollment to capture the manual report source information from the audit table. Kindly review and let me know if there is anything missing.

chrisndodge commented on common/djangoapps/student/models.py:
```diff
@@ -59,6 +59,26 @@
+UN_ENROLL_TO_ENROLL = 'from unenrolled to enrolled'
+ALLOWED_TO_ENROLL_TO_UN_ENROLL = 'from allowed to enroll to enrolled'
+UN_ENROLL_TO_UN_ENROLL = 'from unenrolled to unenrolled'
+DEFAULT_TRANSITION_STATE = 'N/A'
```
If these are strings that get displayed to a user (or administrator), they should be localized.

chrisndodge commented on common/djangoapps/student/models.py:
```diff
@@ -1291,6 +1311,54 @@ def is_verified_enrollment(self):
 
+class ManualEnrollmentAudit(models.Model):
+    """
+    Table for track which enrollments were performed through manual enrollment.
```
spelling: tracking

chrisndodge commented on common/djangoapps/student/models.py:
```diff
@@ -1291,6 +1311,54 @@ def is_verified_enrollment(self):
+    """
+    enrollment = models.ForeignKey(CourseEnrollment, null=True)
+    enrolled_by = models.ForeignKey(User, null=True)
+    enrolled_email = models.CharField(max_length=255, db_index=True)
```
Just to be clear, we use this field for when someone is entered into the AllowedEnrollments table but he/she has not actually gone through the enrollment process. Right?

muhhshoaib replied:
@chrisndodge Once the user is entered into the AllowedEnrollments table, then after that he/she will register/activate his/her account and after account activation we have to add a new entry in the ManualEnrollmentAudit table with the state_transition to ALLOWEDTOENROLL_TO_ENROLL and at that time we don't have information `enrolled_by` and `reason` fields. So we get that information by fetching the latest manual_enrollment_audit entry with the enrolled_email field.

chrisndodge commented on common/djangoapps/student/models.py:
```diff
@@ -1291,6 +1311,54 @@ def is_verified_enrollment(self):
+    @classmethod
+    def get_manual_enrollment_by_email(cls, email):
+        """
+        return manual enrollment by email
```
might want to comment that this returns the most recent entry in the table

chrisndodge commented on common/djangoapps/student/models.py:
```diff
@@ -1291,6 +1311,54 @@ def is_verified_enrollment(self):
+    def get_manual_enrollment(cls, enrollment):
+        """
+        return the manual enrollment object if student enrollment
+        matches else returns None.
```
might want to comment that this returns the most recent entry in the table

chrisndodge commented on common/djangoapps/student/models.py:
```diff
@@ -59,6 +59,26 @@
+    (ALLOWED_TO_ENROLL_TO_UN_ENROLL, ALLOWED_TO_ENROLL_TO_UN_ENROLL),
+    (UN_ENROLL_TO_UN_ENROLL, UN_ENROLL_TO_UN_ENROLL),
+    (DEFAULT_TRANSITION_STATE, DEFAULT_TRANSITION_STATE)
+)
```
I hate to say this because you probably have a number of places to update this, but reading something like ALLOWED_TO_ENROLL_TO_ENROLL is confusing. 

Maybe if it's something more like:

ALLOWEDTOENROLL_TO_ENROLL

As it makes the ALLOWEDTOENROLL to mean a particular state.

Likewise:

ALLOWEDTOENROLL_TO_UNENROLL

Where the 'TO' word separates two states that have no spacing in them?

chrisndodge commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -37,7 +37,15 @@
+  <p>
+    <label for="reason-field-id">
+      ${_("Enter the reason why the students are Enroll or unEnroll manually.")}
+      ${_("The reason field cannot be left blank.")} </label>
```
Can I suggest the following text:

"Enter the reason why the students are to be manually enrolled or unenrolled. This cannot be left blank and will be recorded and presented in Enrollment Reports. Therefore, please given enough detail to account for this action."

chrisndodge commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -37,7 +37,15 @@
+    <label for="reason-field-id">
+      ${_("Enter the reason why the students are Enroll or unEnroll manually.")}
+      ${_("The reason field cannot be left blank.")} </label>
+    <textarea rows="2" id="reason-field-id" name="reason-field" placeholder="${_('Reason')}" spellcheck="false"></textarea>
```
Also, I just noticed on the sandbox that the reason field is not cleared out when the action is submitted. Can you clear this out because otherwise, it is hard to tell that the action actually went through.

chrisndodge commented:
Oops. @muhhshoaib @afzaledx just noticed a blocking bug here. If I do manual enrollment on a non-ecommerce course (that is free) then I get an error message that "Reason cannot be blank". We only need a 'reason' field when the course is not free, so we'll need to add that condition check to the JS.

chrisndodge commented on lms/static/coffee/src/instructor_dashboard/membership.coffee:
```diff
@@ -374,12 +376,18 @@ class BatchEnrollment
+        if not @$reason_field.val()
+          @fail_with_error gettext "Reason field should not be left blank."
+          return false
+
```
Hrm. It seems like you are properly checking if it is a white label course here. But looking at the <input type='hidden' ..> in the HTML, you are just putting the Python boolean which I think will serialize to 'True' and 'False', both of which will always be treated as a true statement in JavaScript. Here if you have to explicitly check to see:

if @$is_course_white_label == 'True'

muhhshoaib commented:
@chrisndodge i have made the changes you suggested.

chrisndodge commented:
@muhhshoaib thanks. I'll refresh the sandbox.

chrisndodge commented:
@muhhshoaib thanks. The immediate bug has been resolved.

muhhshoaib commented:
jenkins, test this please.

muhhshoaib commented:
jenkins, test this please.

chrisndodge commented:
@muhhshoaib for your failing build, looking at the individual shard logs, I see:

10:13:24 Traceback (most recent call last):
10:13:24   File "./manage.py", line 116, in <module>
10:13:24     execute_from_command_line([sys.argv[0]] + django_args)
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 443, in execute_from_command_line
10:13:24     utility.execute()
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 382, in execute
10:13:24     self.fetch_command(subcommand).run_from_argv(self.argv)
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 261, in fetch_command
10:13:24     klass = load_command_class(app_name, subcommand)
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 69, in load_command_class
10:13:24     module = import_module('%s.management.commands.%s' % (app_name, name))
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/utils/importlib.py", line 35, in import_module
10:13:24     __import__(name)
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/south/management/commands/__init__.py", line 10, in <module>
10:13:24     import django.template.loaders.app_directories
10:13:24   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/template/loaders/app_directories.py", line 23, in <module>
10:13:24     raise ImproperlyConfigured('ImportError %s: %s' % (app, e.args[0]))
10:13:24 django.core.exceptions.ImproperlyConfigured: ImportError contentstore: cannot import name user_by_anonymous_id

This doesn't seem related to this PR. @benpatterson @jzoldak are there any generic failures happening now?

@muhhshoaib for now, I'd try get lastest master and rebase this branch (squashing commits) and repushing.

cahrens commented:
Note that migrations need to be load tested-- check in with dev ops about that. @jimabramson can also be a resource, as he has loaded tested a migration.

benpatterson commented:
I have not seen that failure on master. I would suspect something unique about this PR (or, maybe something was cherry-picked?) But that kind of error is not 'flaky'.

chrisndodge commented on common/djangoapps/student/migrations/0050_auto__add_manualenrollmentaudit__chg_field_userprofile_bio.py:
```diff
@@ -0,0 +1,224 @@
+
+
+        # Changing field 'UserProfile.bio'
+        db.alter_column('auth_userprofile', 'bio', self.gf('django.db.models.fields.CharField')(max_length=3000, null=True))
```
@muhhshoaib following up on @cahrens comment on migrations, I'm surprised by the addition of the bio column. Seems like we might have picked that up from somewhere else? Can you investigate.

chrisndodge commented:
@cahrens I brought up your comment about load-testing migrations at standup and it seems like the team is not aware of that step of the PR process. You might want to chat with @smagoun to see if that step of the process is widely socialized.

cahrens commented:
I would suggest that @smagoun talks to DevOps and/or Miki about it. It isn't a TNL requirement per se.

chrisndodge commented:
@muhhshoaib I rebased against current master (to see if that makes a difference with regards to that seemingly unrelated error in the builds). I also regenerated the migration file, but I also see that ALTER statement regarding the 'bio' field length sneaking back into the migration.py file. So I manually edited the migration file.

chrisndodge commented:
@muhhshoaib seems like we now have a successful build. Not sure if the rebase against master fixed it or if it was something else.

cahrens commented:
@muhhshoaib you need to follow up on @chrisndodge 's review comments and get a thumbs-up from him before TNL reviews.

muhhshoaib commented:
@chrisndodge seems like when you squshed the commits, my changes were lost during squshing/rebasing. I had addressed all your feedback suggestions and now it looks like they are not here. 
Luckily that commit was in local repo and i cheery picked that commit and pushed it again.

muhhshoaib commented:
jenkins, test this please.

chrisndodge commented:
sorry about that. There's that bokchoy error again. Can you rebase against master?

muhhshoaib commented:
@chrisndodge rebased with master.

muhhshoaib commented:
jenkins, test this please.

chrisndodge commented:
Hmm. Still getting this exception in the tests:

13:09:41 ========================================
13:09:41  Running tests for cms 
13:09:41 ========================================
13:09:45 Traceback (most recent call last):
13:09:45   File "./manage.py", line 116, in <module>
13:09:45     execute_from_command_line([sys.argv[0]] + django_args)
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 443, in execute_from_command_line
13:09:45     utility.execute()
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 382, in execute
13:09:45     self.fetch_command(subcommand).run_from_argv(self.argv)
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 261, in fetch_command
13:09:45     klass = load_command_class(app_name, subcommand)
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 69, in load_command_class
13:09:45     module = import_module('%s.management.commands.%s' % (app_name, name))
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/utils/importlib.py", line 35, in import_module
13:09:45     __import__(name)
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django_nose/management/commands/test.py", line 10, in <module>
13:09:45     from south.management.commands.test import Command
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/south/management/commands/__init__.py", line 10, in <module>
13:09:45     import django.template.loaders.app_directories
13:09:45   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/template/loaders/app_directories.py", line 23, in <module>
13:09:45     raise ImproperlyConfigured('ImportError %s: %s' % (app, e.args[0]))
13:09:45 ImproperlyConfigured: ImportError contentstore: cannot import name user_by_anonymous_id

This is coming from the CMS unit tests, so we should be able to reproduce it locally. I'll give it a shot.

chrisndodge commented:
Thanks for the fix @afzaledx.

This get's a +1 from me.

@cahrens build failure has been identified and fixed. This is ready for 2nd review. I'll update description to give more details.

cahrens commented:
@chrisndodge OK, please also make sure the sandbox is up-to-date. We will have someone review this on Monday.

dianakhuang commented on common/djangoapps/student/models.py:
```diff
@@ -59,6 +59,26 @@
+ENROLLED_TO_UNENROLLED = _('from enrolled to unenrolled')
+UNENROLLED_TO_ENROLLED = _('from unenrolled to enrolled')
+ALLOWEDTOENROLL_TO_UNENROLLED = _('from allowed to enroll to enrolled')
+UNENROLLED_TO_UNENROLLED = _('from unenrolled to unenrolled')
```
It doesn't seem like a good idea to translate these strings, because it looks like we're never going to be exposing them to users.

If we definitely need these translations, I think these strings will be confusing to translators so I would suggest putting in translators' notes: http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#python-source-code

chrisndodge replied:
These strings will be displayed to instructors who download this report.

cahrens replied:
But will they be based on the language where the server which is generating the report lives, or the instructor's language? I don't think we have translated these sort of things in the past (same sort of issue as translating log messages).

dianakhuang replied:
@chrisndodge we aren't translating [this line](https://github.com/edx/edx-platform/pull/8165/files#diff-9030096cb9659025eaf82ed406dcb128R423) which is also going into the DB, so I don't think these strings should be translated.

dianakhuang commented:
I've finished my quick overall code pass.

I'm also getting 500 errors while trying to [view the courseware](http://manual-enrollment.m.sandbox.edx.org/courses/edX/DemoX/Demo_Course/info)  while on the sandbox. Could someone please check that? Thanks!

chrisndodge commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -413,7 +419,11 @@ def register_and_enroll_students(request, course_id):  # pylint: disable=too-man
                     if not CourseEnrollment.is_enrolled(user, course_id):
-                        CourseEnrollment.enroll(user, course_id)
+                        enrollment_obj = CourseEnrollment.enroll(user, course_id)
+                        reason = 'Enrolling via csv upload'
```
I18N please

chrisndodge commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -427,7 +437,11 @@ def register_and_enroll_students(request, course_id):  # pylint: disable=too-man
                     try:
-                        create_and_enroll_user(email, username, name, country, password, course_id)
+                        enrollment_obj = create_and_enroll_user(email, username, name, country, password, course_id)
+                        reason = 'Enrolling via csv upload'
```
I18N please

chrisndodge commented:
@dianakhuang fixed sandbox. There were some outstanding migrations required. Apologies.

muhhshoaib commented:
jenkins, test this please.

dianakhuang commented:
Checked out the feature on the sandbox. It looks good to me. :+1:

chrisndodge commented:
Rebased and squashed commits. Waiting for build to complete, then I'm merging.

chrisndodge commented:
Jenkins, test this please
